python/ffsim/tenpy/circuits/gates.py

```diff
@@ -0,0 +1,460 @@
+# (C) Copyright IBM 2024.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+import cmath
+import math
+
+import numpy as np
+import tenpy.linalg.np_conserved as npc
+from tenpy.algorithms.tebd import TEBDEngine
+from tenpy.linalg.charges import LegPipe
+from tenpy.networks.mps import MPS
+from tenpy.networks.site import SpinHalfFermionSite
+
+from ffsim.linalg import givens_decomposition
+from ffsim.spin import Spin
+
+# ignore lowercase argument and variable checks to maintain TeNPy naming conventions
+# ruff: noqa: N803, N806
+
+# define sites
+shfs = SpinHalfFermionSite(cons_N="N", cons_Sz="Sz")
+shfsc = LegPipe([shfs.leg, shfs.leg])
+
+
+def sym_cons_basis(gate: np.ndarray) -> np.ndarray:
+    r"""Convert a gate to the TeNPy (N, Sz)-symmetry-conserved basis.
+
+    Args:
+        gate: The quantum gate.
+
+    Returns:
+        The quantum gate in the TeNPy (N, Sz)-symmetry-conserved basis.
+    """
+
+    # convert to (N, Sz)-symmetry-conserved basis
+    if np.shape(gate) == (4, 4):  # 1-site gate
+        swap_list = [1, 3, 0, 2]
+    elif np.shape(gate) == (16, 16):  # 2-site gate
+        swap_list = [5, 11, 2, 7, 12, 15, 9, 14, 1, 6, 0, 3, 8, 13, 4, 10]
+    else:
+        raise ValueError(
+            "only 1-site and 2-site gates implemented for symmetry basis conversion"
+        )
+
+    P = np.zeros(np.shape(gate))
+    for i, s in enumerate(swap_list):
+        P[i, s] = 1
+
+    gate_sym = P.T @ gate @ P
```

Comment:
You don't need to construct the permutation matrix explicitly. Instead, use the inverse permutation,
```python
perm = [10, 8, 2, 11, 14, 0, 9, 3, 12, 6, 15, 1, 4, 13, 7, 5]
```
and do
```python
return gate[perm][:, perm]
```